tests/curve_fit_test.py

```diff
@@ -0,0 +1,507 @@
+    rng = np.random.default_rng(seed)
+    da_coords = {
+        c: sc.linspace(
+            dim=kw.pop('dim', c),
+            **kw,
+        )
+        for c, kw in coords.items()
+    }
+    data = f(**da_coords, **params)
+    # Noise is random but avoiding unbounded values to avoid flaky tests
+    data.values += noise_scale * np.clip(rng.normal(size=data.values.shape), -1.5, 1.5)
+    return sc.DataArray(data, coords=da_coords)
+
+
+def array1d(*, a=1.2, b=1.3, noise_scale=0.1, size=50):
+    return array(
+        dict(x=dict(dim='xx', start=-0.1, stop=4.0, num=size)),
+        func,
+        dict(a=a, b=b),
+        noise_scale=noise_scale,
+    )
+
+
+def array2d(*, a=1.2, b=1.3, noise_scale=0.1, size=20):
+    return array(
+        dict(
+            x=dict(dim='xx', start=-0.1, stop=4.0, num=size),
+            t=dict(dim='tt', start=0.0, stop=1.0, num=size // 2),
+        ),
+        func2d,
+        dict(a=a, b=b),
+        noise_scale=noise_scale,
+    )
+
+
+def array3d(*, a=1.2, b=1.3, noise_scale=0.1, size=10):
+    return array(
+        dict(
+            x=dict(dim='xx', start=-0.1, stop=4.0, num=size),
+            t=dict(dim='tt', start=0.0, stop=1.0, num=size // 2),
+            y=dict(dim='yy', start=2.0, stop=4.0, num=size),
+        ),
+        func3d,
+        dict(a=a, b=b),
+        noise_scale=noise_scale,
+    )
+
+
+def array1d_from_vars(*, a, b, noise_scale=0.1, size=50):
+    return array(
+        dict(x=dict(dim='xx', start=0.1, stop=4.0, num=size, unit='m')),
+        func,
+        dict(a=a, b=b),
+        noise_scale=noise_scale,
+    )
+
+
+def test_should_not_raise_given_function_with_dimensionless_params_and_1d_array():
+    curve_fit(['x'], func, array1d())
+
+
+@pytest.mark.parametrize(
+    "p0, params, bounds",
+    (
+        (None, dict(a=1.2, b=1.3), None),
+        (dict(a=3, b=2), dict(a=1.2, b=1.3), None),
+        (dict(a=0.2, b=-1), dict(a=1.2, b=1.3), {'a': (-3, 3), 'b': (-2, 2)}),
+        (dict(a=0.2, b=-1), dict(a=1.2, b=1.3), {'a': (-3, 1.1), 'b': (-2, 1.1)}),
+    ),
+)
+@pytest.mark.parametrize(
+    "dims",
+    (
+        dict(x=10, t=10, y=10),
+        dict(x=5, t=8, y=7),
+    ),
+)
+@pytest.mark.parametrize(
+    "coords, reduce_dims",
+    (
+        (['x'], []),
+        (['x'], ['y']),
+        (['x'], ['t', 'y']),
+        (['x', 't'], []),
+        (['x', 't'], ['y']),
+        (['x', 't', 'y'], []),
+    ),
+)
+def test_compare_to_curve_fit_xarray(dims, coords, reduce_dims, p0, params, bounds):
+    f, fxarray = {
+        1: (func, func_np),
+        2: (func2d, lambda x, a, b: func2d_np(*x, a, b)),
+        3: (func3d, lambda x, a, b: func3d_np(*x, a, b)),
+    }[len(coords)]
+    da = array(
+        {c: dict(start=1, stop=3, num=n) for c, n in dims.items()},
+        lambda **x: (
+            sc.broadcast(
+                f(**{c: x[c] for c in x if c in coords or c in params}),
+                sizes=dims,
+            ).copy()
+        ),
```

Comment:
keyword args! It is completely unclear what this does otherwise.

What is this function? Why is there a broadcast?

Reply:
I'll add kwargs. The `array` helper takes a `func` argument that creates the data of the array from the coords.
If the function has fewer arguments than the array has dimensions then we need to broadcast it to the right shape.

Although, I realized now that if we just use `func3d` in all cases then the result will always have the right shape anyway and it's much easier to read.

Reply:
I think broadcasting might hide a category of potential bugs, so using `func3d` sounds better.